awx/api/serializers.py

```diff
@@ -4293,6 +4293,29 @@ def __init__(self, *args, **kwargs):
                 error_list.append(_("Field '{}' unavailable".format(exc.message)))
             except SecurityError as exc:
                 error_list.append(_("Security error due to field '{}'".format(exc.message)))
+
+        # Ensure that if a webhook body was provided, that it can be rendered as a dictionary
+        notification_type = ''
+        if self.instance:
+            notification_type = getattr(self.instance, 'notification_type', '')
+        else:
+            notification_type = self.initial_data.get('notification_type', '')
+
+        if notification_type == 'webhook':
+            for event in messages:
+                body = messages[event].get('body', {})
+                if body:
+                    try:
+                        # !!! DANGER ZONE: Evaluating user input !!!
+                        # Remove globals/locals entirely, then attempt to render the dictionary
+                        potential_body = eval(body, {"__builtins__": None}, {})
```

Comment:
Nope, whatever use case this serves, I'd rather just not support it if it means having to use `eval` here with user input.

Reply:
on second thought, seems like I could just use `json.loads`

Reply:
```python
>>> s = "[x for x in 'x'.__class__.__bases__[0].__subclasses__() if x.__name__ == 'Popen'][0](['whoami'])"
>>> eval(s, {'__builtins__': None}, {})
<subprocess.Popen object at 0x7f540965c128>
>>> ryanpetrello
```

😉

Reply:
Fun (related) reading: https://nedbatchelder.com/blog/201206/eval_really_is_dangerous.html

Reply:
> The problem with all of these attempts to protect eval() is that they are blacklists. They explicitly remove things that could be dangerous. That is a losing battle because if there’s just one item left off the list, you can attack the system.

☝️ 

good read, danke